packages/expo-notifications/android/src/main/java/expo/modules/notifications/tokens/PushTokenManager.java

```diff
@@ -0,0 +1,104 @@
+package expo.modules.notifications.tokens;
+
+import org.unimodules.core.interfaces.SingletonModule;
+
+import java.lang.ref.WeakReference;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.WeakHashMap;
+
+import expo.modules.notifications.tokens.interfaces.PushTokenListener;
+
+public class PushTokenManager implements SingletonModule, expo.modules.notifications.tokens.interfaces.PushTokenManager {
+  private static final String SINGLETON_NAME = "PushTokenManager";
+
+  /**
+   * We store this value to be able to inform new listeners of last known token.
+   */
+  private String mLastToken;
+
+  /**
+   * A collection of weak references to listeners. Used to iterate over, on new token.
+   */
+  private Collection<WeakReference<PushTokenListener>> mListeners;
+
+  /**
+   * A weak map of listeners -> reference. Lets us check quickly whether given listener
+   * is already registered.
+   */
+  private WeakHashMap<PushTokenListener, WeakReference<PushTokenListener>> mListenerReferenceMap;
```

Comment:
Same as in `FirebaseTokenListenerService`, wouldn't map be just enough and `mListeners` redundant?